ostd/src/mm/page_table/cursor.rs

```diff
@@ -734,26 +734,97 @@ where
+    /// All the mappings in the current cursor's range must be empty. The
+    /// function allows the source cursor to operate on the mapping before
+    /// the copy happens. So it is equivalent to protect then duplicate.
+    /// Only the mapping is copied, the mapped pages are not copied.
+    ///
+    /// It can only copy tracked mappings since we consider the untracked
+    /// mappings not useful to be copied.
+    ///
+    /// After the operation, both cursors will advance by the specified length.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure that
+    ///  - the range being copied with the operation does not affect kernel's
+    ///    memory safety.
+    ///  - both of the cursors are in tracked mappings.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if:
+    ///  - either one of the range to be copied is out of the range where any
+    ///    of the cursor is required to operate;
+    ///  - either one of the specified virtual address ranges only covers a
+    ///    part of a page.
+    ///  - the current cursor's range contains mapped pages.
+    pub unsafe fn copy_from(
+        &mut self,
+        src: &mut Self,
+        len: usize,
+        op: &mut impl FnMut(&mut PageProperty),
+    ) {
+        assert!(len % page_size::<C>(1) == 0);
+        let this_end = self.0.va + len;
+        assert!(this_end <= self.0.barrier_va.end);
+        let src_end = src.0.va + len;
+        assert!(src_end <= src.0.barrier_va.end);
 
-        while self.0.level < C::NR_LEVELS {
-            self.0.level_up();
-        }
+        while self.0.va < this_end && src.0.va < src_end {
+            let cur_pte = src.0.read_cur_pte();
+            if !cur_pte.is_present() {
+                src.0.move_forward();
+                continue;
+            }
 
-        self.0.guards[(C::NR_LEVELS - 1) as usize].take()
+            // Go down if it's not a last node.
+            if !cur_pte.is_last(src.0.level) {
+                src.0.level_down();
+
+                // We have got down a level. If there's no mapped PTEs in
+                // the current node, we can go back and skip to save time.
+                if src.0.guards[(src.0.level - 1) as usize]
+                    .as_ref()
+                    .unwrap()
+                    .nr_children()
+                    == 0
+                {
+                    src.0.level_up();
+                    src.0.move_forward();
+                }
 
-        // Ok to drop the cursor here because we ensure not to access the page table if the current
-        // level is the root level when running the dropping method.
+                continue;
+            }
+
+            // Do protection.
+            let mut pte_prop = cur_pte.prop();
+            op(&mut pte_prop);
+
+            let idx = src.0.cur_idx();
+            src.cur_node_mut().protect(idx, pte_prop);
+
+            // Do copy.
+            let child = src.cur_node_mut().child(idx, true);
```

Comment:
In `src.cur_node_mut().protect()` and  `src.cur_node_mut().child()`, there are two additional `read_pte()` opeartions since we have got the cur_pte before. 

Is there any way to avoid them ?( If there isn't a elegant way, that's fine, since the additional overhead isn't significant). For example, we can add an additional API like this:
```rust
    {
     ...
        cur_pte.set_prop(pte_prop);
        src.cur_node_mut().write_pte(idx, cur_pte);
        let child = src.cur_node_mut().extract_pte_info(cur_pte, true);
    ...
    }

    pub(super) fn extract_pte_info(&self, pte: &E, in_tracked_range: bool) -> Child<E, C> {
        if !pte.is_present() {
            Child::None
        } else {
            let paddr = pte.paddr();
            if !pte.is_last(self.level()) {
                // SAFETY: We have a reference count to the page and can safely increase the reference
                // count by one more.
                unsafe {
                    Page::<PageTablePageMeta<E, C>>::inc_ref_count(paddr);
                }
                Child::PageTable(RawPageTableNode {
                    raw: paddr,
                    _phantom: PhantomData,
                })
            } else if in_tracked_range {
                // SAFETY: We have a reference count to the page and can safely
                // increase the reference count by one more.
                unsafe {
                    DynPage::inc_ref_count(paddr);
                }
                // SAFETY: The physical address of the PTE points to a forgotten
                // page. It is reclaimed only once.
                Child::Page(unsafe { DynPage::from_raw(paddr) }, pte.prop())
            } else {
                Child::Untracked(paddr, pte.prop())
            }
        }
    }

    pub(super) fn child(&self, idx: usize, in_tracked_range: bool) -> Child<E, C> {
        debug_assert!(idx < nr_subpage_per_huge::<C>());

        let pte = self.read_pte(idx);
        self.extract_pte_info(&pte, in_tracked_range)
    }
```

Reply:
Good direction, but I have another plan.

Reply:
I think it touches some fundamental safety issues of the page table node design. So I want to postpone this optimization, which should save a few memory reads.